flink-connector-aws/flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/metrics/KinesisSourceShardMetrics.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.connector.kinesis.source.metrics;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.annotation.VisibleForTesting;
+import org.apache.flink.connector.kinesis.source.split.KinesisShardSplit;
+import org.apache.flink.connector.kinesis.source.split.KinesisShardSplitState;
+import org.apache.flink.metrics.MetricGroup;
+import org.apache.flink.runtime.metrics.groups.AbstractMetricGroup;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import software.amazon.awssdk.arns.Arn;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * A utility class for handling metrics in {@link
+ * org.apache.flink.connector.kinesis.source.reader.KinesisStreamsSourceReader}.
+ */
+@Internal
+public class KinesisSourceShardMetrics {
+    private static final Logger LOG = LoggerFactory.getLogger(KinesisSourceShardMetrics.class);
+
+    private final Map<String, Long> millisBehindLatestMap = new HashMap<>();
+    private final Map<String, MetricGroup> shardMetricGroupMap = new HashMap<>();
+    private final MetricGroup sourceReaderMetricGroup;
+
+    public KinesisSourceShardMetrics(final MetricGroup metricGroup) {
+        this.sourceReaderMetricGroup = metricGroup;
+    }
+
+    public void registerShardMetricGroup(KinesisShardSplit shard) {
```

Comment:
Do we want to do a check to validate that the shard is not already registered?

Reply:
Added the check.